### DISCUSSION TITLE
TS error in defineConfig: 'test' does not exist in type 'UserConfigExport'

### DISCUSSION THREAD

**Discussion by misha-erm** [vitest-dev/vitest#1106] [Q&A] [answered]
Hello,

I have some problems with setting up vitest to work with vite and ts...

Here is my vite.config,ts
<img width="499" alt="image" src="https://user-images.githubusercontent.com/8783498/161816247-3d20f98a-4bc4-44a9-9cf5-2af54ca68e60.png">

I added reference types as mentioned in https://vitest.dev/config/#configuration but it didn't help.

Here are my tsconfigs:

tsconfig.json

```
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "module": "ESNext",
    "moduleResolution": "Node",
    "noEmit": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ESNext",
    "types": ["vite/client", "vitest/globals"],
    "allowJs": true,

    "baseUrl": "./",
  },
  "include": ["./src"],
  "references": [{ "path": "./tsconfig.node.json" }]
}

```
tsconfig.node.json
```
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Node",
    "composite": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "skipLibCheck": true,
    "strict": true
  },
  "include": ["vite.config.ts", "tests"]
}
```

Any ideas? Thank you

**Comment by james-elicx** [comment, marked as answer]
I had the same problem and defining it like this with an arrow function fixed the issue.

```ts
export default defineConfig(() => ({
  build: {
    // ...
  },
  test: {
    // ...
  },
}));
```

**Reply by misha-erm** [reply]
Yep, yesterday found this solution too. Using function works for me

**Reply by ferdicus** [reply]
this won't give you autocompletion for the types though... this just gets rid of the error message 😕

**Reply by UsamaEjaz0** [reply]
Can someone explain why this works?

**Reply by jeffvandyke** [reply]
@UsamaEjaz0 This works because defineConfig accepts both a UserConfig and a `type UserConfigFnObject = (env: ConfigEnv) => UserConfig;`, and a function type definition requires only that the returned object minimally satisfy the declared return type, and won't complain if extra properties are present. This is meant to make it easier to reuse a multi-purpose function in a callback that only requires a minimum satisfying type. It's a bit of a workaround here because we _should_ know the full object type, and not allow unknown props, but that's a bit tricky with the "test" config here.

**Reply by sagarpanchal** [reply]
This is a bad solution. Check https://github.com/vitest-dev/vitest/discussions/1106#discussioncomment-11863078

**Reply by mltsy** [reply]
Using a function works okay if you only need to define one config.  But we have multiple configs, and use mergeConfig to merge them and apply overrides.  mergeConfig unfortunately, does not allow UserConfigFnObject as an argument, so I'm still stuck without a solution.

**Reply by sagarpanchal** [reply]
> Using a function works okay if you only need to define one config. But we have multiple configs, and use mergeConfig to merge them and apply overrides. mergeConfig unfortunately, does not allow UserConfigFnObject as an argument, so I'm still stuck without a solution.

https://github.com/vitest-dev/vitest/discussions/1106#discussioncomment-11863078

**Comment by james-elicx** [comment]
Looking at one of my other projects, the following reference at the top of your config file will also fix the issue, instead of needing to use a function.

```ts
/// <reference types="vitest/config" />
```

**Reply by AntonOfTheWoods** [reply]
But also make sure you have compatible versions of `vite` and `vitest`.

**Reply by maczyt** [reply]
Yes, `latest vite` should use `vitest 3.x`

**Reply by sagarpanchal** [reply]
This is incomplete, so it might not work. If this does not work. https://github.com/vitest-dev/vitest/discussions/1106#discussioncomment-11863078

**Comment by SiNONiMiTY** [comment]
Just merge the UserConfig interface & InlineConfig interface and get the full type safety 🚀

```ts
import { defineConfig, type UserConfig } from 'vite'
import type { InlineConfig } from 'vitest/node'

export default defineConfig({
    build: {
        ...
    },
    test: {
        ...
    },
} as UserConfig & { test: InlineConfig })
```

**Reply by sagarpanchal** [reply]
You are using `as` to suppress type-checking, not type safe. https://github.com/vitest-dev/vitest/discussions/1106#discussioncomment-11863078

**Comment by sagarpanchal** [comment]
This issue occurs because of the versions of the `vite`, `vitest` and `@vitest/**` packages are incompatible.

If you're using latest `vite` then use the latest `vitest` and latest `@vitest/**` packages. 

- Add the reference in the `vite.config.**` file
```ts
/// <reference types="vitest/config" />
```

- Install all latest versions or all matching version
```sh
npm i -D vite@latest vitest@latest @vitest/utils@latest @vitest/coverage-v8@latest @vitest/ui@latest
```

**OR**

- Use [ncu](https://www.npmjs.com/package/npm-check-updates) to select and update all vite and vitest packages
```sh
ncu -i --format group
```

And, for immediate effect, restart the ts server in your IDE or restart the IDE.

**Reply by aqeelat** [reply]
the reference file only worked for me when I was using .js. Syncing the versions worked but it gets a drift very frequently (I don't know why). I'm using bun so this might be a local caching thing. I gave up and used two files.
